doc: TRD: start non-XIP flash info document (#4081)
tock/tock: doc/reference/trd-non-xip.md

Tock on Platforms Without Execute-in-Place (XIP) Flash
========================================

**TRD:** <br/>
**Working Group:** Kernel<br/>
**Type:** Informational<br/>
**Status:** Draft <br/>
**Author:** Brad Campbell<br/>
**Draft-Created:** 2024/07/05<br/>
**Draft-Modified:** 2024/07/05<br/>
**Draft-Version:** 1<br/>
**Draft-Discuss:** devel@lists.tockos.org<br/>

Abstract
-------------------------------

This document provides information relating to running Tock on platforms without
execute-in-place flash storage.

Introduction
===============================

Traditional embedded systems use execute-in-place (XIP) flash to both store
program code and to enable executing that program code. However, newer hardware

> designs, particularly those in security-focused settings, do not include any XIP
> flash storage. These platforms, subsequently referred to as "non-xip", only

Reviewer: ```suggestion
designs, particularly those integrated into a system-on-chip alongside an application processor, do not include 
executable persistent storage. These platforms, subsequently referred to as "non-xip", can only
```

> execute instructions stored in volatile RAM. All executable code must be stored
> in external nonvolatile storage, loaded into RAM, and then executed.

Reviewer: ```suggestion
execute instructions stored in volatile RAM. All executable code must be loaded
from external nonvolatile storage, loaded into RAM, and then executed.
```

Author: Is it not stored anywhere?

Author: This was very confusing to me during the discussion, the code has to exist _somewhere_, no? Or is part of the assumption that the machine code may actually be completely lost at some point?

Reviewer: it's both true that it needs to be stored somewhere and that it needs to be loaded from somewhere. Between the two, I just think that the more salient part is that Tock needs to load it from somewhere.

Author: I think it is helpful to clarify that the assumption is the code is stored somewhere, and even if it is evicted from RAM it could be retrieved again. If that assumption is _not_ expected to be true I think that could very materially change potential Tock designs.

Reviewer 2: I think that we shouldn't unnecessarily constrain ourselves by saying that Tock always has access to this code, which to me is implied by "stored". I think we want to support use-cases where a kernel has ephemeral access to binaries, or they are "streamed" over some interface (e.g., downloaded over TCP into memory). I think "loaded" encapsulates this a little better.

Author: In that case, that seems like a significant variation that we need to consider explicitly (and why this document exists).

Reviewer 2: I'm not sure whether we talked about this _explicitly_ in the breakout session during TW7, but it felt implied. For instance, applications may be provided in storage that's "leased" from some larger chip where that memory is either not always available, or may be modified at any given time (for our purposes, volatile external memory). To support these use-cases it seems like it's vital for Tock's support to _not_ require unconditional access to such external storage.

But it'd be great if others can confirm whether or not that's a real concern. @lawrenceesswood @reynoldsbd

Author: I just want to replace the implication with something concretely written down one way or the other because I think my assumptions have been generally different from the others in that breakout.

Reviewer 3: At work I often work on non-XIP MCU platforms, and one common use-case during development is the ability to stream code directly into RAM via JTAG and then execute from RAM, without having to store that code in flash at any intermediate point. This has several advantages:
1. it is faster
2. it reduces flash write cycles, which can be useful for hardware that is running hourly CI runs
3. if you have less flash than RAM, this enables running debug builds that only fit in RAM but do not fit in flash. This can be highly useful if your application is flash constrained in practice but you occasionally want to be able to run true debug builds (i.e. unoptimized code)

Of course, Tock could deign not to support this mode of operation. 

It is also possible on these platforms for OTA updates to overwrite the flash that contains the currently running application (since that application has already been copied into RAM). This is another common use case where Tock assuming that the application running in RAM matches the application that exists in flash could be problematic.

Reviewer 4: f22a2900bbd0c831d964837543328bd4f650c50c

This system architecture shrinks the amount of the hardware that must be trusted
when securely executing programs. By not including execute-in-place flash, the
flash storage does not need to be trusted to correctly store program code and
data. When the program is copied to RAM the program binary can be verified
(i.e., by checking a cryptographic signature) before the loaded binary is
allowed to execute.

2.0 Tradeoffs
===============================

Non-xip platforms introduce a new design point for embedded devices with a
different set of tradeoffs compared to traditional platforms. To help motivate
potential Tock designs for non-xip platforms, we discuss both the limitations
compared to traditional platforms, and the properties that may be important on
traditional platforms but are likely not a priority on non-xip platforms.

Limitations:

- Non-xip platforms have less RAM that traditional platforms have XIP flash.
  Further, this RAM on non-xip platforms must (at runtime) store both the
  executable code and the conventional memory items (e.g., the stack and heap).
  This results in significantly less room for program code than on traditional

>   platforms.

Reviewer 3: This is not universally true -- for example, TI sells the popular AM263x / AM263Px line of chips which have 2MB and 3MB of RAM, respectively.

Reviewer: ```suggestion
- Non-xip platforms store (at runtime) both the
  executable code and the conventional memory items (e.g., data, stack and heap).
  This results in significantly less room for program code than on traditional
  platforms.
```

Reviewer: Can confirm that understanding of one of the real use cases in question has order of MBs of RAM (another has 100s of KB so, again, both are correct).

> - As flash storage is not trusted, any data retrieved from flash storage must be
>   verified before it is used.

Reviewer: ```suggestion
- As flash storage may not be trustworthy, code retrieved from flash storage may need to be
  verified before it is used.
```

Non-limitations:

- Energy is not a bottleneck, and additional power draw from having to load

>   program code into RAM is acceptable.

Reviewer 4: This may be too strong of a statement. IIUC, low-power operation is of interest and a pending priority for cases where Tock is an always-on component for battery-powered devices, e.g. laptops, phones.

That said, the energy overhead of additional RAM use may not be a significant factor; though, I think there was also some discussion of chips with banked SRAM and the capacity to power off unused portions, so it may yet be significant.

Reviewer: True, though also to add to the list of things that may not be an important consideration in this context (but yes in others) is the energy overhead of copying memory back and forth between external/persistent storage and RAM, or the energy overhead of swapping things out.

Perhaps

```suggestion
- Energy is not as significant a bottleneck as in other settings, and additional power draw from having to load
  program code into RAM is acceptable.
```

Reviewer 4: I actually have a half-baked research project with some preliminary data from a thing a MS student started for a quarter if you want to run down that alley further...

As newer MCUs integrate more interesting non-volatile storage options (hello MRAM on the Apollo4), the energy tradeoff gets much tighter. To a first order, for the Apollo4 it's around 1.5~2m where the tipping point occurs where it's cheaper in energy to write a byte out to MRAM and read it back than to preserve it in SRAM — for processes that wake less often, swapping them out and disabling banks of SRAM could save energy, at least on paper :).

Reviewer 4: [trd-non-xip 9c7147b19] trd-non-xip: energy nuance comment

- Execution time has relaxed deadlines, and additional execution time from
  having to load program code into RAM is acceptable.

> 

Reviewer: ```suggestion
- Loading code into RAM allows more predictable layout of code and data and/or rewriting some instructions during load time, potentially enabling simpler relocation and shared libraries.
```

>   having to load program code into RAM is acceptable.

Reviewer 4: Does this need to be qualified on "For some applications"?

Author: These non-xip platforms have to give up _something_ right? They can't possibly be _as_ low power and _as_ fast with 25%-50% of the code storage as a conventional MCU can they? I can't imagine we can develop something if we can't assume some tradeoffs in all non-xip cases.

Reviewer 4: Yeah, I think what I was probing at here was that _some_ (and likely few) applications could be time sensitive and would need to perhaps be 'pinned' or similar, which is another design point to consider.

Reviewer 3: On many of these platforms program code is only loaded into RAM after reset, so that is the only time there is an execution delay. Otherwise, these platforms often run faster than XIP platforms, as SRAM accesses are generally faster than NOR flash accesses.

The need to "page in" applications from flash at runtime is definitely not universal to all non-XIP MCU platforms -- just those with significantly less flash than SRAM.

Reviewer: Yes, though @hudson-ayers, two specific use cases that came for discussions are those in which there actually is not enough SRAM to concurrently have loaded all the processes that one might want. E.g., TPM2 is just too darn big to fit with more than one or two other processes on a reasonably sized platform.

In one use case, swapping was mentioned as an option, in another, just killing apps on eviction would be acceptable, but both cases involve loading/unloading from SRAM at runtime.

And in both of these cases, I think _some_ applications would prefer to be pinned.

Basically both sides seem to be true here.

Reviewer 5: Dynamically loading parts of apps is definitely an important use case. Tock [already supports that](https://github.com/tock/tock/blob/master/kernel/src/platform/platform.rs#L262) :)

Reviewer 4: [trd-non-xip 6b19cbbc3] trd-non-xip: capture dynamic app load nuances
